### PR TITLE
Add latest version of MPFR and patches for older versions

### LINKED ISSUE #7543: gcc v5.4.0 build fails due to mpfr patching problem
There seems to be a patch application issue in the mpfr-3.1.5 build procedure

I was expecting something like my previous build:
```
==> Installing mpfr
==> Fetching file://MIRROR_DIR/mirror/mpfr/mpfr-3.1.5.tar.bz2
==> Staging archive: WORKING_DIR/var/spack/stage/mpfr-3.1.5-rmi7bmi3oaqduvjown2v46snr6ps2zr5/mpfr-3.1.5.tar.bz2
==> Created stage in WORKING_DIR/var/spack/stage/mpfr-3.1.5-rmi7bmi3oaqduvjown2v46snr6ps2zr5
==> Applied patch vasprintf.patch
==> Applied patch strtofr.patch
==> Building mpfr [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Executing phase: 'build'
==> Executing phase: 'install'
==> Successfully installed mpfr
  Fetch: 0.04s.  Build: 9.54s.  Total: 9.58s.
[+] WORKING_DIR/opt/spack/linux-centos7-x86_64/gcc-4.8.5/mpfr-3.1.5-rmi7bmi3oaqduvjown2v46snr6ps2zr5
```
When I tried to build the gcc compiler yesterday (and again this morning) the results were strange:
```
==> Installing mpfr
1 out of 1 hunk FAILED -- saving rejects to file VERSION.rej
1 out of 1 hunk FAILED -- saving rejects to file src/mpfr.h.rej
1 out of 1 hunk FAILED -- saving rejects to file src/version.c.rej
==> Fetching file://MIRROR_DIR/mirror/mpfr/mpfr-3.1.5.tar.bz2
==> Staging archive: WORKING_DIR/sat/spack/var/spack/stage/mpfr-3.1.5-rmi7bmi3oaqduvjown2v46snr6ps2zr5/mpfr-3.1.5.tar.bz2
==> Created stage in WORKING_DIR/sat/spack/var/spack/stage/mpfr-3.1.5-rmi7bmi3oaqduvjown2v46snr6ps2zr5
==> Patch strtofr.patch failed.
==> Error: ProcessError: Command exited with status 1:
    '/usr/bin/patch' '-s' '-p' '1' '-i' 'WORKING_DIR/sat/spack/var/spack/repos/builtin/packages/mpfr/strtofr.patch' '-d' '.'
==> Error: [Errno 2] No such file or directory: 'WORKING_DIR/sat/spack/var/spack/stage/mpfr-3.1.5-rmi7bmi3oaqduvjown2v46snr6ps2zr5/mpfr-3.1.5/spack-build.out'
```
Not only the error, but the order of the messages seem strange.

A clean clone of the spack repo made no difference
```console
$ spack install gcc@5.4.0
```

Default environment:
```linux-centos7-x86_64/gcc-4.8.5```

### PR DESCRIPTION
**Pull request by adamjstewart**
Fixes #7543. @peetsv 

Adds the latest version of MPFR, which builds and passes all tests for me on macOS 10.13.3 with Clang 9.0.0.

The MPFR developers are some of the few people I've seen who actually release patches to fix bugs in older versions of software. They release these patches in two forms: individual patches and cumulative patches. I was using individual patches in #3506, but after #7193 was merged, the order in which patches are applied is no longer the order in which they are listed in the package, causing them to conflict (see #7543). This PR replaces the individual patches with cumulative patches and adds patches for older releases as well.

Note that some of these patches are fairly large. Spack does support fetching patches from the internet, but I don't think I'm comfortable fetching patches remotely without checksumming them. And since the cumulative patches are updated continuously as bugs are discovered, checksumming them would just lead to headaches.

Pinging @alalazo and @davydden who have had opinions on long patches in the past.

### CODE REVIEW

**Comment by davydden** [comment]
> Note that some of these patches are fairly large. Spack does support fetching patches from the internet, but I don't think I'm comfortable fetching patches remotely without checksumming them. And since the cumulative patches are updated continuously as bugs are discovered, checksumming them would just lead to headaches.

I don't have anything against this. We fetch it in some scenarios but also store them locally in Spack for other packages.

**Comment by peetsv** [comment]
Consistency (the dubious bugaboo) would suggest that having patch files stored with tarballs in a local mirror would not be too out of place.  That said, checksum support of those files would also add some fun.

As a complete non sequitur, I would like to have a spack function to populate my local mirror based upon the package definitions - source tarballs and patches too if that's the way this gets played.   So, assuming your mirrors.yaml file is defined somewhere ...
```./bin/spack mirror populate perl@5.24.1```

yeah, I know, "Go away kid, ya bother me"

**Comment by adamjstewart** [comment]
> I would like to have a spack function to populate my local mirror based upon the package definitions

You mean like `spack mirror create`? http://spack.readthedocs.io/en/latest/mirrors.html

**Comment by peetsv** [comment]
Almost.  I already have a mirror that was established long ago.  What I really need is a simple way to add things as package prerequisites change...  Maybe a flag or something that tells spack to put a copy of anything it downloads into the local mirror...

**Comment by alalazo** [comment]
> Spack does support fetching patches from the internet, but I don't think I'm comfortable fetching patches remotely without checksumming them.

You can checksum them, according to the directive signature:
```python
def patch(url_or_filename, level=1, when=None, working_dir=".", **kwargs):
    """Packages can declare patches to apply to source.  You can
    optionally provide a when spec to indicate that a particular
    patch should only be applied when the package's spec meets
    certain conditions (e.g. a particular version).

    Args:
        url_or_filename (str): url or filename of the patch
        level (int): patch level (as in the patch shell command)
        when (Spec): optional anonymous spec that specifies when to apply
            the patch
        working_dir (str): dir to change to before applying

    Keyword Args:
        sha256 (str): sha256 sum of the patch, used to verify the patch
            (only required for URL patches)
        archive_sha256 (str): sha256 sum of the *archive*, if the patch
            is compressed (only required for compressed URL patches)

    """
    ....
```

**Comment by alalazo** [comment]
> I would like to have a spack function to populate my local mirror based upon the package definitions

@adamjstewart @peetsv 
OT: I think there's another small feature that has been missing for a long time, i.e. the possibility to refer to a mirror by name. Right now you can do:
```console
$ spack mirror create -d <DIR> <specs>
```
but given that when you add a mirror you name it, I'd like to be able to say also:
```console
$ spack mirror create -name <MIRROR_NAME> <specs>
```

**Comment by adamjstewart** [comment]
> > Spack does support fetching patches from the internet, but I don't think I'm comfortable fetching patches remotely without checksumming them.
>
> You can checksum them, according to the directive signature:

I know, but:

> since the cumulative patches are updated continuously as bugs are discovered, checksumming them would just lead to headaches.

Do you think it would be worth fetching anyway? The package would break every time a bug fix came out. I don't want to deal with that many issues.

**Comment by alalazo** [comment]
I left a comment in #7193. Personally I'd prefer fixing the sorting  and continue with individual patches. There are other packages, like [nwchem](http://www.nwchem-sw.org/index.php/Download), for which patching order might be important, e.g. if they don't provide cumulative patches.

**Comment by adamjstewart** [comment]
I agree that we should fix patching order. But there are 5-10 individual patches per version of this package, so that may lead to clutter.

**Comment by adamjstewart** [comment]
@alalazo I was thinking something along the same lines. I just checked and it looks like they don't release patches for old releases after a new release comes out. So it should be safe to fetch and checksum patches for old releases. For the current release, we could either fetch individual patches or store cumulative patches. There are no patches for the current release so far, so we can dodge this question for now.

I still vote we use cumulative patches wherever possible. Version 3.1.3 for example includes 17 individual patches.

File: var/spack/repos/builtin/packages/mpfr/package.py [hunk 1]
```diff
@@ -30,18 +30,33 @@ class Mpfr(AutotoolsPackage):
...
+    # http://www.mpfr.org/mpfr-X.Y.Z/#bugs
+    patches = {
+        '3.1.6': '66a5d58364113a21405fc53f4a48f4e8',
+        '3.1.5': '1dc5fe65feb5607b89fe0f410d53b627',
+        '3.1.4': 'd124381573404fe83654c7d5a79aeabf',
+        '3.1.3': 'ebd1d835e0ae2fd8a9339210ccd1d0a8',
+        '3.1.2': '9f96a5c7cac1d6cd983ed9cf7d997074',
+    }
+
+    for version, checksum in patches.items():
```
[the remark this answers is not in the document]
Comment (alalazo) [reply var/spack/repos/builtin/packages/mpfr/package.py]: Using:
```python
for v, c in patches.items():
    ...
```
should help avoiding troubles, as I am pretty sure we don't define those attributes in base classes (joy and pain of dynamic typing).

**Comment by alalazo** [comment]
> 1 and 2 don't make any sense to me, but they remind me of #3129. I suspect that the same bug is causing both of these problems.

:clap: for remembering things like that more than one year later :sweat_smile:

File: var/spack/repos/builtin/packages/mpfr/package.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (alalazo) [reply var/spack/repos/builtin/packages/mpfr/package.py]: In #7563 I ended up adding a double underscore to loop variables. Not sure what's the worst choice between very short names and names that starts with a double underscore. Should we set an informal style policy for the few packages that have logic at class level?

[the remark this answers is not in the document]
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/mpfr/package.py]: How silly of me, fixed in the latest commit. I wonder if there's any way we can make this idiot proof.

[the remark this answers is not in the document]
Comment (alalazo) [reply var/spack/repos/builtin/packages/mpfr/package.py]: > I wonder if there's any way we can make this idiot proof.

Maybe using a descriptor and intercepting the redefinition at class level. Not sure it's worth the effort though.

**Pull request merged**